tests/async/test_dispatch_event.py

```diff
@@ -0,0 +1,174 @@
+
+
+async def test_should_dispatch_click_after_a_cross_origin_navigation(page, server):
+    await page.goto(server.PREFIX + "/input/button.html")
+    await page.dispatchEvent("button", "click")
+    await page.goto(server.CROSS_PROCESS_PREFIX + "/input/button.html")
+    await page.dispatchEvent("button", "click")
+    assert await page.evaluate("() => result") == "Clicked"
+
+
+async def test_should_not_fail_when_element_is_blocked_on_hover(page, server):
+    await page.setContent(
+        """<style>
+      container { display: block; position: relative; width: 200px; height: 50px; }
+      div, button { position: absolute; left: 0; top: 0; bottom: 0; right: 0; }
+      div { pointer-events: none; }
+      container:hover div { pointer-events: auto; background: red; }
+    </style>
+    <container>
+      <button onclick="window.clicked=true">Click me</button>
+      <div></div>
+    </container>"""
+    )
+    await page.dispatchEvent("button", "click")
+    assert await page.evaluate("() => window.clicked")
+
+
+async def test_should_dispatch_click_when_node_is_added_in_shadow_dom(page, server):
+    await page.goto(server.EMPTY_PAGE)
+    watchdog = page.dispatchEvent("span", "click")
+    await page.evaluate(
+        """() => {
+      const div = document.createElement('div');
+      div.attachShadow({mode: 'open'});
+      document.body.appendChild(div);
+    }"""
+    )
+    await page.evaluate("() => new Promise(f => setTimeout(f, 100))")
+    await page.evaluate(
+        """() => {
+      const span = document.createElement('span');
+      span.textContent = 'Hello from shadow';
+      span.addEventListener('click', () => window.clicked = true);
+      document.querySelector('div').shadowRoot.appendChild(span);
+    }"""
+    )
+    await watchdog
+    assert await page.evaluate("() => window.clicked")
+
+
+async def test_should_be_atomic(selectors, page, utils):
+    await utils.register_selector_engine(
+        selectors,
+        "dispatchEvent",
+        """{
+            create(root, target) { },
+            query(root, selector) {
+                const result = root.querySelector(selector);
+                if (result)
+                Promise.resolve().then(() => result.onclick = "");
+                return result;
+            },
+            queryAll(root, selector) {
+                const result = Array.from(root.querySelectorAll(selector));
+                for (const e of result)
+                Promise.resolve().then(() => result.onclick = "");
+                return result;
+            }
+        }""",
+    )
+    await page.setContent('<div onclick="window._clicked=true">Hello</div>')
+    await page.dispatchEvent("dispatchEvent=div", "click")
+    assert await page.evaluate("() => window._clicked")
+
+
+@pytest.mark.skip("TODO: Determine why there is a failure.")
+@pytest.mark.skip_browser("webkit")
+async def test_should_dispatch_drag_drop_events(page, server):
```

Comment:
This test and the below test are failing hence marked as skip; I have not yet looked into the reason for failure. The test executes but the final assertion fails. I suspect it probably has to with `dataTransfer`

Reply:
Looking at prints from dispatchEvent:

```
print(eventInit)
print(serialize_argument(eventInit))
```

I'm seeing:

```
{<playwright.async_api.JSHandle object at 0x10ea51c50>}
{'value': {'v': 'undefined'}, 'handles': []}
```

So I'm definitely not serializing the JSHandle properly before sending it over the channel.

Reply:
I fixed the accidental set notation, but still seeing a failure to serialize the handle:

```
{'dataTransfer': <playwright.async_api.JSHandle object at 0x10b108d68>}
{'value': {'o': [{'k': 'dataTransfer', 'v': {'v': 'undefined'}}]}, 'handles': []}
```

Stepping into the serialization code now to see why

```
<class 'playwright.async_api.JSHandle'>
```

is not matching the JSHandle check.

Reply:
@pavelfeldman What do you expect `print(await page.evaluateHandle("() => new DataTransfer()"))` to be? Trying to narrow down whether I have the correct thing that's not getting serialized in the eventInit serialization.

Reply:
Great finding!

As a workaround you can use for now (add it before the isinstance):
```py
    if type(value).__name__ == "JSHandle" and not isinstance(value, JSHandle):
        value = value._impl_obj
```
In the meantime we'll see if we can find a better solution for that.

Reply:
There is no DataTransfer constructor in WebKit and that's why these tests are marked as failing upstream with `it.fail(WEBKIT)`.  Does this answer your question of there is something I'm not following?

Reply:
Ah, you are saying it does not work on Chromium also. Looking!

Reply:
@rwoll, so I am late to the party, Max already suggested something. We generate sync and async API layers on top of the core implementation in Python and `isinstance` is checking for the internal type, while user passes the wrapper. The solution is to traverse all the incoming dictionaries recursively and translate wrapper types into regular types. I'll send a PR for that!

Reply:
Also, for debugging, run `DEBUGP=1 pytest --browser=chromium -s` to see the protocol log

Reply:
Fixed in https://github.com/microsoft/playwright-python/pull/136/files

Reply:
> We generate sync and async API layers on top of the core implementation in Python and isinstance is checking for the internal type, while user passes the wrapper.

This is what I was tripping on. Couldn't quite make sense of the fully qualified type `playwright.async_api.JSHandle` when I didn't see it in the source. Thanks both for the suggested workaround and the more thorough fix. 

@pavelfeldman Once you MR lands I'll rebase and remove the more hacky workaround. 🎉